sdk/servicebus/service-bus/src/serviceBusMessage.ts

```diff
@@ -1055,8 +973,20 @@ export class ServiceBusMessageImpl implements ReceivedMessageWithLock {
    * @throws MessagingError if the service returns an error while renewing message lock.
    */
   async renewLock(): Promise<Date> {
-    this.throwIfMessageCannotBeSettled(this.getReceiverFromContext(), "renew the lock on");
-    return await this._context.managementClient!.renewLock(this.lockToken!);
+    if (this.sessionId) {
+      throw translate({
+        description: `Invalid operation on the message, message lock doesn't exist when dealing with sessions`,
+        condition: ErrorNameConditionMapper.InvalidOperationError
+      });
+    } else {
+      const isDeferredMessage = this._context.requestResponseLockedMessages.has(this.lockToken!);
+      const receiver = isDeferredMessage
+        ? undefined
+        : this._context.getReceiver(this.delivery.link.name, this.sessionId);
+      if (!isDeferredMessage) this.throwIfMessageCannotBeSettled(receiver, "renew the lock on");
```

Comment:
Since when did we start using `throwIfMessageCannotBeSettled` for the renewLock operation?
Do we have to do these checks?
Can we not rely on `managementClient.renewLock()` to fail if the lock cannot be renewed?

Reply:
This was added last month as part of preview-1. 
I have not added this but I'd like to defend its presence on the grounds that we have the opportunity to throw the error if something is not right and we don't have to make a service call.

![image](https://user-images.githubusercontent.com/10452642/80072993-de552080-84fb-11ea-92da-6d33bf13fe9c.png)

Reply:
My concerns are around the misdirections when we read this code a few months later
- trying to renew locks, but relying on a method whose name suggests its about settling
- knowing that renewing locks uses $management link, but fetching receiver link and passing it to a validation helper

My recommendation:
- Move the contents of `throwIfMessageCannotBeSettled()` inside `settleMessage()`
- Copy only the relevant parts of `throwIfMessageCannotBeSettled()` into `renewLock()`

Reply:
Side note: Look into history to see why we chose to add the `Failed to .. the message as this message is already settled` instead of relying on the service to return the error. Messages settled using the $management link will not have the `remote_settled` property set to true. So this is not a guarantee that we will throw this error for settled messages

Reply:
> Side note: Look into history to see why we chose to add the Failed to .. the message as this message is already settled instead of relying on the service to return the error.

Added in here - https://github.com/Azure/azure-sdk-for-js/pull/7847  (..Not much info on why it was chosen. @richardpark-msft, you can chime in). 
I'd say, we have the chance to throw an error before making the service call, so no harm in doing it.

> Messages settled using the $management link will not have the remote_settled property set to true. So this is not a guarantee that we will throw this error for settled messages

Not a guarantee but a decent guard. If the remote_settled=true, we know that the message had been settled for sure. 


Updated the code as per your inputs, please review again.